### PR TITLE
Plane: Move navigation functions into flight mode classes.

### PR DESCRIPTION
**Pull request by samuelctabor**
After discussing with @magicrub this should be a step in the right direction. I'm planning to flight test this soon.

### CODE REVIEW

**Comment by magicrub** [comment]
This needs a rebase but otherwise is a straightforward refactor. Sadly, refactors like this don't last long in master without a conflict so lets take a look at this one again after the rebase and get it in. It looks like a low-risk refactor, no behavior change.

**Comment by samuelctabor** [comment]
Good stuff, I will rebase tomorrow. All tests pass on my local branch.

File: ArduPlane/mode_avoidADSB.cpp
Comment (peterbarker) [inline ArduPlane/mode_avoidADSB.cpp:16]: I'm not sure these things should be calling into mode_loiter to do stuff.  They *could* call update_loiter themselves.

On the Copter side we've often factored out the functionality - `land_run_vertical` / `land_run_horizontal`, for example.  That's not universal - for example, mode-auto uses mode-rtl for RTL'ing and mode-guided for guided'ing....

It's a minor point and doesn't need to be addressed, really.

File: ArduPlane/mode_rtl.cpp
Comment (peterbarker) [inline ArduPlane/mode_rtl.cpp:56]: Don't need the else here.

Comment (samuelctabor) [reply ArduPlane/mode_rtl.cpp]: Fixed

**Comment by samuelctabor** [comment]
Rebased and removed the extra "else".

**Comment by samuelctabor** [comment]
Any other changes required before this can be merged or taken to the dev call?

**Comment by peterbarker** [comment]
@samuelctabor fine as far as I can see.  Needs to be merged by a Plane maintainer IMO, 'though - and that's not me :-)

I'll mark it DevCall so you have a deadline at least.

File: ArduPlane/ArduPlane.cpp
Comment (WickedShell) [inline ArduPlane/ArduPlane.cpp:442]: Why retain this wrapper at all? It's only called once from `Plane::navigate` so this is adding an indirection that isn't needed, and makes it harder to follow?

**Comment by rmackay9** [comment]
As discussed on the dev call, @magicrub and @tridge will do a detailed inspection of this before merging.

File: ArduPlane/mode_avoidADSB.cpp
Comment (tridge) [inline ArduPlane/mode_avoidADSB.cpp:16]: I'd like them to call plane.update_loiter() instead. The problem with calling into mode_loiter is it means we're calling a member function on a class wirthout ever calling the init function on the class. Right now this is safe, but if someone changes mode_loiter it could break other modes.
In general I don't like calls into non-static functions on a mode class without the class enter function having been called

**Comment by tridge** [comment]
I've checked for copy&paste errors and it looks good. I'd like someone to test all affected flight modes to look for correct behaviour in a manual SITL run

File: ArduPlane/mode_avoidADSB.cpp
Comment (rmackay9) [reply ArduPlane/mode_avoidADSB.cpp]: yes, I totally agree with Tridge's comment.

Comment (samuelctabor) [reply ArduPlane/mode_avoidADSB.cpp]: Changed to call into plane.update_loiter().

**Comment by samuelctabor** [comment]
Checked:
 - correct circle radius and location in takeoff, guided and loiter
 - auto mission works
 - flying around in cruise works
 - RTL works correctly with various RTL_AUTOLAND settings

ToDo:
 - having some trouble with avoid ADSB, I'll check with Tom tomorrow
 - RTL with quadplane

**Comment by samuelctabor** [comment]
- RTL with quadplane works correctly
 - ADSB avoidance has other [issues](https://github.com/ArduPilot/ardupilot/issues/15043) but this PR doesn't change them

**Comment by samuelctabor** [comment]
Perhaps we could discuss this on the EU dev call?

**Comment by rmackay9** [comment]
Added the DevCallEU at @samuelctabor's request.

File: ArduPlane/ArduPlane.cpp
Comment (tridge) [reply ArduPlane/ArduPlane.cpp:442]: I agree, it is clearer without it

**Review by tridge — CHANGES_REQUESTED** [review]
just remove the extra level of indirection, so plane.update_loiter() is called directly by the mode navigate functions, then can be merged when passes CI

**Comment by peterbarker** [comment]
Rebased this on top of CI problem fixes.

**Comment by samuelctabor** [comment]
Thanks Peter, I did the same a few seconds later before I saw your comment :-)

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**
